File: problems/Q3/Q3.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const n = s.length
    let charmap = {}

    let res = 0
    let i = 0

    for (const j of Array(n).keys()) {
        i = Math.max(charmap[s[j]] || -1, i)
        res = Math.max(res, j - i + 1)
        charmap[s[j]] = j + 1
    }
    return res
};

console.log(lengthOfLongestSubstring(
    "abcabc"
));

File: problems/Q3/Q3.lua
function lengthOfLongestSubstring( s --[[string]] )
    local charmap = {}

    local res = 0
    local i = 0

    for j = 1, #s do
        local char = s:sub(j, j)
        if charmap[char] ~= nil then
            i = math.max( charmap[char], i)
        end
        res = math.max(res, j - i)
        charmap[char] = j
    end
    return res
end

print(lengthOfLongestSubstring(
    "abcabc"
))